drivers/usb/udc/udc_mcux.c

```diff
@@ -0,0 +1,903 @@
+	case USB_EP_TYPE_BULK:
+		ep_init.transferType = USB_ENDPOINT_BULK;
+		break;
+	case USB_EP_TYPE_INTERRUPT:
+		ep_init.transferType = USB_ENDPOINT_INTERRUPT;
+		break;
+	case USB_EP_TYPE_ISO:
+		ep_init.transferType = USB_ENDPOINT_ISOCHRONOUS;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return usb_mcux_control(dev, kUSB_DeviceControlEndpointInit, &ep_init);
+}
+
+static int usb_mcux_ep_disable(const struct device *dev,
+			       struct udc_ep_config *const cfg)
+{
+	LOG_DBG("Disable ep 0x%02x", cfg->addr);
+
+	return usb_mcux_control(dev, kUSB_DeviceControlEndpointDeinit, &cfg->addr);
+}
+
+static int usb_mcux_host_wakeup(const struct device *dev)
+{
+	return -ENOTSUP;
+}
+
+static int usb_mcux_set_address(const struct device *dev, const uint8_t addr)
+{
+	uint8_t temp_addr = addr;
+
+	return usb_mcux_control(dev, kUSB_DeviceControlSetDeviceAddress, &temp_addr);
+}
+
+static int usb_mcux_enable(const struct device *dev)
+{
+	return usb_mcux_control(dev, kUSB_DeviceControlRun, NULL);
+}
+
+static int usb_mcux_disable(const struct device *dev)
+{
+	return usb_mcux_control(dev, kUSB_DeviceControlStop, NULL);
+}
+
+static int usb_mcux_init(const struct device *dev)
+{
+	const struct udc_mcux_config *config = dev->config;
+	struct udc_mcux_data *priv = udc_get_private(dev);
+	usb_status_t status;
+	uint8_t index;
+
+#ifdef CONFIG_USB_UDC_NXP_PHY
+	usb_phy_config_struct_t phy_config;
+#endif
+
+#ifdef CONFIG_USB_UDC_NXP_EHCI
+#if defined(USBHS_STACK_BASE_ADDRS)
+	uint32_t usb_base_addrs[] = USBHS_STACK_BASE_ADDRS;
+#else
+	uint32_t usb_base_addrs[] = USBHS_BASE_ADDRS;
+#endif
+#endif
+#ifdef CONFIG_USB_UDC_NXP_IP3511
+#ifdef USB_BASE_ADDRS
+	uint32_t ip3511_fs_base[] = USB_BASE_ADDRS;
+#endif
+#ifdef USBHSD_BASE_ADDRS
+	uint32_t ip3511_hs_base[] = USBHSD_BASE_ADDRS;
+#endif
+#endif
+
+	/* get the right controller id */
+	priv->controller_id = 0xFFu; /* invalid value */
+#ifdef CONFIG_USB_UDC_NXP_EHCI
+	for (index = 0; index < ARRAY_SIZE(usb_base_addrs); index++) {
+		if (usb_base_addrs[index] == config->base) {
+			priv->controller_id = kUSB_ControllerEhci0 + index;
+			break;
+		}
+	}
+#endif
+#ifdef CONFIG_USB_UDC_NXP_IP3511
+#ifdef USB_BASE_ADDRS
+	for (index = 0; index < ARRAY_SIZE(ip3511_fs_base); index++) {
+		if (ip3511_fs_base[index] == config->base) {
+			priv->controller_id = kUSB_ControllerLpcIp3511Fs0 + index;
+			break;
+		}
+	}
+#endif
+
+#ifdef USBHSD_BASE_ADDRS
+	if (priv->controller_id == 0xFF) {
+		for (index = 0; index < ARRAY_SIZE(ip3511_hs_base); index++) {
+			if (ip3511_hs_base[index] == config->base) {
+				priv->controller_id = kUSB_ControllerLpcIp3511Hs0 + index;
+				break;
+			}
+		}
+	}
+#endif
+#endif
+	if (priv->controller_id == 0xFFu) {
+		return -ENOMEM;
+	}
+
+#ifdef CONFIG_USB_UDC_NXP_PHY
+	if (config->need_to_init_phy) {
+		phy_config.D_CAL = config->D_CAL;
+		phy_config.TXCAL45DP = config->TXCAL45DP;
+		phy_config.TXCAL45DM = config->TXCAL45DM;
+		USB_EhciPhyInit(priv->controller_id, 0u, &phy_config);
```

Comment:
Why is `usb_phy_config_struct_t phy_config;` not a part of `udc_mcux_config`?

Reply:
Updated.